python/paddle/fluid/tests/unittests/dygraph_to_static/test_tensor_shape.py

```diff
@@ -58,17 +58,116 @@ def dyfunc_tensor_shape_5(x):
     return res
 
 
-test_funcs = [
-    dyfunc_tensor_shape_1, dyfunc_tensor_shape_2, dyfunc_tensor_shape_3,
-    dyfunc_tensor_shape_4, dyfunc_tensor_shape_5
-]
+def dyfunc_with_if_1(x):
+    x = fluid.dygraph.to_variable(x)
+    res = fluid.layers.reshape(x, [-1, 1])
+    x_shape_0 = x.shape[0]
+    if x_shape_0 < 1:
+        # `res.shape[0] > 1` is transformed into `if fluid.layers.shape(res)[0] > 1`
+        if res.shape[0] > 1:
+            res = fluid.layers.fill_constant(
+                value=2, shape=x.shape, dtype="int32")
+        else:
+            res = fluid.layers.fill_constant(
+                value=3, shape=x.shape, dtype="int32")
+    return res
+
+
+def dyfunc_with_if_2(x):
+    x = fluid.dygraph.to_variable(x)
+    # `len(x.shape)` will not be transformed.
+    if len(x.shape) < 1:
+        res = x
+    else:
+        res = fluid.layers.fill_constant(value=8, shape=x.shape, dtype="int32")
+
+    return res
+
+
+def dyfunc_with_for_1(x):
+    x = fluid.dygraph.to_variable(x)
+    res = fluid.layers.fill_constant(value=0, shape=[1], dtype="int32")
+    # `x.shape[0]` is transformed into `fluid.layers.shape(x)[0]`
+    for i in range(x.shape[0]):
+        res += 1
+    return res
+
+
+def dyfunc_with_for_2(x):
+    x = fluid.dygraph.to_variable(x)
+    x_shape_0 = x.shape[0]
+    res = fluid.layers.fill_constant(value=0, shape=[1], dtype="int32")
+
+    # `x_shape_0` is transformed into `fluid.layers.shape(x)[0]`
+    for i in range(x_shape_0):
+        res += 1
+    return res
+
+
+def dyfunc_with_for_3(x):
+    # TODO(liym27):
+    #  It will fail to run because `for i in range(len(x.shape))` will be transformed into Paddle while_loop.
+    #  After the AST tranformation of for loop is improved, add TestTensorShapeInFor3.
+    x = fluid.dygraph.to_variable(x)
+    res = fluid.layers.fill_constant(value=0, shape=[1], dtype="int32")
+    # `len(x.shape)` is not transformed.
+    for i in range(len(x.shape)):
+        res += 1
+
+    return res
+
+
+def dyfunc_with_while_1(x):
+    x = fluid.dygraph.to_variable(x)
+    res = fluid.layers.fill_constant(value=0, shape=[1], dtype="int32")
+    # `x.shape[0]` is transformed into `fluid.layers.shape(x)[0]`
+    i = 1
+    while i < x.shape[0]:
+        res += 1
+        i = i + 2
+    return res
+
+
+def dyfunc_with_while_2(x):
+    x = fluid.dygraph.to_variable(x)
+    x_shape_0 = x.shape[0]
+    res = fluid.layers.fill_constant(value=0, shape=[1], dtype="int32")
+    i = 1
+    # `x_shape_0` is transformed into `fluid.layers.shape(x)[0]`
+    # TODO(liym27): If `x_shape_0` is at right like `while i < x_shape_0`, it will not be transformed.
+    #  Fix this bug next PR.
+    while x_shape_0 > i:
+        res += 1
+        i = i + 2
+    return res
 
 
-class TestTensorShape(unittest.TestCase):
+def dyfunc_with_while_3(x):
+    # TODO(liym27):
+    #  It will fail to run because `while len(x_shape) < i` will be transformed into Paddle while_loop and `x_shape` will be loop_vars.
```

Comment:
Is here still error? I think x_shape should be one of loop vars, because it is in the condition

Reply:
Same as above. x_shape is a python list. But here the while-loop shouldn't be transformed into `while_loop`.